### ISSUE TITLE
Revisions Blocks not populated

### ISSUE THREAD

**Issue by lnesi** [pristas-peter/wp-graphql-gutenberg#21]
On revisions, blocks do not get resolved data when you use saveContent the data comes in 

See example:
```
{
  "data": {
    "pageBy": {
      "title": "Test Page 2",
      "pageId": 544,
      "revisions": {
        "nodes": [
          {
            "content": "\n<div class=\"wp-block-cover has-background-dim\" style=\"background-image:url(http://localhost:8080/app/uploads/2019/11/1000-24.jpg)\"><div class=\"wp-block-cover__inner-container\">\n<p style=\"text-align:center\" class=\"has-large-font-size\">Test Page edited HI</p>\n</div></div>\n\n\n\n\n\n\n\n<p></p>\n",
            "blocks": []
          },
          {
            "content": "\n<div class=\"wp-block-cover has-background-dim\" style=\"background-image:url(http://localhost:8080/app/uploads/2019/11/1000-24.jpg)\"><div class=\"wp-block-cover__inner-container\">\n<p style=\"text-align:center\" class=\"has-large-font-size\">Test Page edited sdd</p>\n</div></div>\n\n\n\n\n\n\n\n<p></p>\n",
            "blocks": []
          },
          {
            "content": "\n<div class=\"wp-block-cover has-background-dim\" style=\"background-image:url(http://localhost:8080/app/uploads/2019/11/1000-24.jpg)\"><div class=\"wp-block-cover__inner-container\">\n<p style=\"text-align:center\" class=\"has-large-font-size\">Test Page edited 2</p>\n</div></div>\n\n\n\n\n\n\n\n<p></p>\n",
            "blocks": []
          },
          {
            "content": "\n<div class=\"wp-block-cover has-background-dim\" style=\"background-image:url(http://localhost:8080/app/uploads/2019/11/1000-24.jpg)\"><div class=\"wp-block-cover__inner-container\">\n<p style=\"text-align:center\" class=\"has-large-font-size\">Test Page edited</p>\n</div></div>\n\n\n\n\n\n\n\n<p></p>\n",
            "blocks": []
          },
          {
            "content": "\n<div class=\"wp-block-cover has-background-dim\" style=\"background-image:url(http://localhost:8080/app/uploads/2019/11/1000-24.jpg)\"><div class=\"wp-block-cover__inner-container\">\n<p style=\"text-align:center\" class=\"has-large-font-size\">Test Page</p>\n</div></div>\n\n\n\n\n\n\n\n<p></p>\n",
            "blocks": []
          },
          {
            "content": "\n<div class=\"wp-block-cover has-background-dim\" style=\"background-image:url(http://localhost:8080/app/uploads/2019/11/1000-24.jpg)\"><div class=\"wp-block-cover__inner-container\">\n<p style=\"text-align:center\" class=\"has-large-font-size\">Test Page</p>\n</div></div>\n\n\n\n\n\n\n\n<p></p>\n",
            "blocks": []
          },
          {
            "content": "\n<div class=\"wp-block-cover has-background-dim\" style=\"background-image:url(http://localhost:8080/app/uploads/2019/11/1000-24.jpg)\"><div class=\"wp-block-cover__inner-container\">\n<p style=\"text-align:center\" class=\"has-large-font-size\">Test Page</p>\n</div></div>\n\n\n\n\n\n\n\n<p></p>\n",
            "blocks": []
          },
          {
            "content": "\n<div class=\"wp-block-cover has-background-dim\" style=\"background-image:url(http://localhost:8080/app/uploads/2019/11/1000-24.jpg)\"><div class=\"wp-block-cover__inner-container\">\n<p style=\"text-align:center\" class=\"has-large-font-size\">Test Page</p>\n</div></div>\n\n\n\n\n\n\n\n<p></p>\n",
            "blocks": []
          },
          {
            "content": "\n<div class=\"wp-block-cover has-background-dim\" style=\"background-image:url(http://localhost:8080/app/uploads/2019/11/1000-24.jpg)\"><div class=\"wp-block-cover__inner-container\">\n<p style=\"text-align:center\" class=\"has-large-font-size\">Test Page</p>\n</div></div>\n\n\n\n\n\n\n\n<p></p>\n",
            "blocks": []
          },
          {
            "content": "\n<div class=\"wp-block-cover has-background-dim\" style=\"background-image:url(http://localhost:8080/app/uploads/2019/11/1000-24.jpg)\"><div class=\"wp-block-cover__inner-container\">\n<p style=\"text-align:center\" class=\"has-large-font-size\">Test Page</p>\n</div></div>\n\n\n\n\n\n\n\n<p></p>\n",
            "blocks": []
          }
        ]
      },
      "blocks": [
        {
          "saveContent": "<div class=\"wp-block-cover has-background-dim\" style=\"background-image:url(http://localhost:8080/app/uploads/2019/11/1000-24.jpg)\"><div class=\"wp-block-cover__inner-container\"><!-- wp:paragraph {\"align\":\"center\",\"placeholder\":\"Write title…\",\"fontSize\":\"large\"} -->\n<p style=\"text-align:center\" class=\"has-large-font-size\">Test Page edited sdd</p>\n<!-- /wp:paragraph --></div></div>",
          "name": "core/cover"
        },
        {
          "saveContent": "",
          "name": "lndev-blocks/post-list-by-category"
        },
        {
          "saveContent": "",
          "name": "lndev-blocks/post-list-by-category"
        },
        {
          "saveContent": "<p></p>",
          "name": "core/paragraph"
        }
      ]
    }
  }
}
```
query
```
# Try to write your query here
query {
  pageBy(pageId: 544) {
    title
    pageId
    revisions{
      nodes{
        content
        blocks{
          name
          saveContent
        }
      }
    }
    blocks{
      saveContent
      name
          
    }
  }
}
```

**Comment by janfeldmann** [comment]
Any news on this? Would be great to have when building the previews.

**Comment by pristas-peter** [comment]
Hi guys, sorry I am busy with gatsby-wordpress-gutenberg however i will get back to this plugin after im done with that. However supporting revisions wont be an easy task, because of how gutenberg works, however if  you are using gatsby the issue would be resolved in the latter plugin.

**Comment by janfeldmann** [comment]
Hey @pristas-peter 

in which way will the issue be resolved by gatsby?

Greetings Jan

**Comment by pristas-peter** [comment]
I am writting a completely new native gatsby plugin which will work without this plugin. The development is done here: https://github.com/pristas-peter/gatsby-wordpress-gutenberg

**Comment by mrclay** [comment]
@pristas-peter Could you give a high level overview of the fix in case one of us could attempt it?

**Comment by pristas-peter** [comment]
Sure, you would need to query all revisions inside gutenberg (js) and parse all postContent with window.wp.blocks api. Then send it to php on post save (just how it is done now) and then register it with graphql

**Comment by janfeldmann** [comment]
@pristas-peter 
As I understand it, you use the "update_callback" from the REST API to use the update_post_meta function.

This is not triggered, when I just hit the "Preview" button without saving the post right?

Do you know if I can use another hook here?
For example if the post autosaves are updated?

**Comment by janfeldmann** [comment]
Hey, I think I did it. 
I will do create a pull request when I have some spare time.

Basically I registered a new field
register_rest_field(..., wp_graphql_gutenberg_preview, ...)
for the REST API and the meta data

Inside /src/blocks.js I added a **isPreviewUpdateRequest** function to check if the user hit the preview button. Inside I post the data to the REST API

`wp.apiFetch({
	path: `/wp/v2/${restBase}/${post.id}`,
	method: "PUT",
	data: {
		wp_graphql_gutenberg_preview: data
	}
});`

I access this data inside 'update_callback' from the newly register_rest_field from above and update the 'wp_graphql_gutenberg_preview' post meta data.

**Comment by pristas-peter** [comment]
Yes, fantastic, PR is welcomed. 
If you want to complete this, theres on more feature needed in admin js, when the force update is taking place, because this will only work for newly saved posts. 
 Thanks

**Comment by mrclay** [comment]
@janfeldmann Any progress here? I'd welcome partial functionality to allow post previews to be able to display.

**Comment by Splact** [comment]
Following @janfeldmann solution I created this pull request #26 .

Don't know if it's a solution valid to be merged since it's much cleaner having blocks data accessible directly from revision. Here instead we have a blocksPreview field (that contains the blocks data coming from the latest autosave) on the post object. Anyway this solved my needs, so I think it's worth sharing.

So when you need to query preview data, your query will be something like:

```graphql
query ($wpId: Int!) {
  customPostExample: customPostExampleBy(customPostExampleId: $wpId) {
    title
    slug
    content
    # use good old `blocks` as fallback when no autosave was fired
    blocks {
      name
      ... on AcfExampleBlock {
        acf {
          test
          test_2
        }
      }
      ... on CoreParagraphBlock {
        originalContent
      }
    }
    blocksPreview {
      name
      ... on AcfExampleBlock {
        acf {
          test
          test_2
        }
      }
      ... on CoreParagraphBlock {
        originalContent
      }
    }
    
    preview: revisions(first: 1, before: null) {
      nodes {
        title
        slug
        content
      }
    }
  }
}

```

**Comment by pristas-peter** [comment]
@lnesi @kingkero @gustavpursche @janfeldmann @mrclay 

Hi @Splact, thanks for the PR, however i think it is an temporary solution and we should address this in a larger scale.

I was wondering, would you guys mind running this plugin rewritten in JS? What I mean you would have to host your own node application (next.js, heroku) or maybe even serverless function (netlify, aws, zeit now) which wp-graphql would proxy to? It would solve a lot of problems like revision, ability to query blocks without re-saving them and so?

**Comment by mrclay** [comment]
@pristas-peter What data would live in this separate app?

**Comment by pristas-peter** [comment]
None, it would act as a sort of provider, parser for wo-graphql

**Comment by pristas-peter** [comment]
Hi guys neverming the conversation above. Fixed in 9bae117

I have rewritten the plugin and included support for revisions and also for live previews as you type (you need to be authenticated and have edit_others_posts capabality to be able to query) like example above:

```graphql
{
  posts {
    edges {
      node {
        blocks {
          __typename
          ... on CoreParagraphBlock {
            attributes {
              ... on CoreParagraphBlockAttributes {
                content
              }
            }
          }
        }
        revisions {
          edges {
            node {
              databaseId
              blocks {
                ... on CoreParagraphBlock {
                  attributes {
                    ... on CoreParagraphBlockAttributes {
                      content
                    }
                  }
                }
              }
            }
          }
        }
        previewBlocks {
          __typename
          ... on CoreParagraphBlock {
            attributes {
              ... on CoreParagraphBlockAttributes {
                content
              }
            }
          }
        }
      }
    }
  }
}
```


Will be included in a new version (0.2.0) introducing some breaking changes. I am in a process of writting new docs. You can test it out from develop branch right now if you want.

**Comment by pristas-peter** [comment]
Released in 0.2.0
